src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1077,28 +1077,31 @@
                            result-binding (conj result-binding)
                            fact-binding (conj fact-binding))
 
-            ;; Find children that all parent nodes have.
-            common-children-ids (into
-                                 ;; Since we require that id-to-condition-nodes have an equal value to "node" under the ID
-                                 ;; for this to be used, the order of this set probably doesn't matter in most cases.  However, in any possible edge
-                                 ;; cases where there are equal nodes under different IDs, sorting the set will add determinism.
-                                 ;; Having different nodes under the same
-                                 ;; ID would be a bug, but having an equivalent node under multiple IDs wouldn't necessarily be one.
-                                 (sorted-set)
-                                 (apply s/union (-> (:forward-edges beta-graph)
-                                                    (select-keys parent-ids)
-                                                    (vals))))
-
+            ; Find children that all parent nodes have.
+            forward-edges (apply s/union (-> (:forward-edges beta-graph)
+                                                (select-keys parent-ids)
+                                                (vals)))
```

Comment:
If I'm reading this logic correctly, isn't an alternative implementation something like:
```clj

node-id (if (seq forward-edges)
          (apply min forward-edges)
          (create-id-fn))
```

Something like that seems to have less building involved than the `sorted-set` or this new `transient {}` impl. I may be missing something about what else is happening here though.

Reply:
only if the "forward edge"'s node is equivalent to that of the node that was constructed

Reply:
Good point. So revising my question, what about:
```clj

node-id (or (some->> forward-edges
                     seq
                     (filter #(= node (get id-to-condition-nodes %)))
                     (apply min))
            (create-id-fn))
```

I'm basically just trying to see if there isn't some alternative that doesn't require such heavy construction of intermediate structures. I guess the `sorted-set` was somewhat expensive. I wonder why? 
Also, `clojure.core/min` is pretty slow I believe. And having to use `apply` since it is variatic is even worse. Could write a faster one (that takes a collection) for this usage and it may make something like my example here even more performant to compare with.

Reply:
From VM snapshots, sorted-set was expensive due to `cons` on the underlying data struct.

The reason i avoided something like the `filter`, was because the `equiv` gets hot, hotter than a simple hash while using `assoc!`.

Reply:
I see, so trying to offload `equiv` by using more hashing. Interesting.
I guess the transient map thing seems reasonable. They do build pretty fast compared to other persistent colls.

The only remaining question I'd have then, is why not just use a sorted collection via `sort` instead of a `sorted-set` to solve the problem.

It might be a bit better to build the `forward-edges` set inline during the `reduce` via `transduce`. I think you'd still want to use `distinct` though, so you don't get to get fully away from building a set.
```clj
node->id (transduce (comp cat (distinct))
                    (completing update-node->id)
                    (-> (:forward-edges beta-graph)
                        (select-keys parent-ids)
                        vals))
```
Maybe not a big deal though if you don't see any reason to question the perf of the `apply s/union` happening here. If `sorted-set` was expensive though, this must be a fairly hot spot. I'm not sure `apply s/union` would be that cheap depending on how many collections were involved typically.

Reply:
As I understand this if we set aside the effort spent on constructing the sorted-set, we have the following sources of algorithmic complexity in a big-O sense:


Ethan's proposal:
````
update-node->id (fn [m id]
                  (let [node (get id-to-condition-nodes id)
                        prev (get m node)]
                    (if prev
                      (assoc! m node (min prev id))
                      (assoc! m node id))))

node->id (reduce update-node->id
                 (transient {})
                 forward-edges)
````
For each forward edge, we'll have a get on an integer key, a get on a node key, a map update and potentially a min/max comparison.  The map gets will be O(n) in the number of elements in forward-edges.

Current version:
````
;; Find an id for an equivalent node.
existing-id (first (for [child-id forward-edges
                         :when (= node (get id-to-condition-nodes child-id))]
                     child-id))
````
Note that I replaced "common-children-ids" with "forward-edges" here for clarity since they have the same information in both versions, albeit with different names.  We basically have O(n) equality comparisons and gets on integer keys in the number of elements in forward-edges.

Between these, it seems like equality could be faster due to potentially quick returns of false and the avoidance of calculating hashcodes on larger objects like nodes.. but I could also mistaken.  The real problem here looks like it is probably the sorted-set, and VisualVM snapshots show lots of time spent in clojure.lang.PersistentTreeSet.cons.  Sorted sets (and sorted maps) don't support transients per [CLJ-955](https://dev.clojure.org/jira/browse/CLJ-995), and I verified this in a REPL.

````
my-ns.core> (transient (sorted-set))
ClassCastException clojure.lang.PersistentTreeSet cannot be cast to clojure.lang.IEditableCollection  clojure.core/transient (core.clj:3233)
````

As a result Clojure will have to resort to building the sorted-set one element at a time in "into"; see [here](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core.clj#L6821).  This will be slow:

````
(into (sorted-set) stuff-here)
````

Was something like the following considered?

````
(sort
  (apply s/union (-> (:forward-edges beta-graph)
                     (select-keys parent-ids)
                     (vals))))
````

The reason I wonder if this might perform better is that the [sort](https://github.com/clojure/clojure/blob/clojure-1.9.0/src/clj/clojure/core.clj#L3107) implementation is ultimately implemented by a sort of a mutable Java array, which is probably pretty performant, even if it isn't of linear complexity.

Regarding the time saved, my understanding is that the actual case motivating this is significantly larger than the example given in the issue; please correct me if I'm mistaken @EthanEChristian.

Reply:
@WilliamParker 
I also considered `sort`. Probably without `s/union`. I don't think `apply s/union` is very fast at all. In particular if it goes into the > 2 arity cases (which I think happens here fairly easily).

Something like
```clj
common-children-ids (->> (select-keys (:forward-edges beta-graph))
                         vals
                         (into [] (comp cat (distinct)))
                         sort)
```

Also,
> Between these, it seems like equality could be faster due to potentially quick returns of false and the avoidance of calculating hashcodes on larger objects like nodes..

Equality may not be quick to return false. Perhaps nodes are always similar in the first keys looked at. However, hash also has its own penalty, as you've mentioned. The hash code may be cached though and perhaps the same nodes are repeatedly hashed. Then again, if they are records, they don't cache hash codes at the top-level at least, until clj v1.9. Even if the top-level is cached though, larger internal structures may be (if they are clj persistent structures for example).

Reply:
Correct, the issue i was seeing has at max ~10k forward edges, however there are multiple calls where the forward edges approach 10k.

These are pretty big sessions, 40k productions.

@mrrodriguez,
Testing the `transduce` suggested seems to be worse performing, do to what seems to be a performance enhancement in the `clojure.set` library when [unioning]( https://github.com/clojure/clojure/blob/master/src/clj/clojure/set.clj#L13-L31) several sets.

Reply:
@EthanEChristian I saw that optimization and did not believe it could actually be very fast. Looks pretty slow. It just picks the biggest set to put in front of the merge. However, the steps to do that seem costly. Also, I didn't think our set sizes should be drastically different in this union op here. However, I don't know that. I'll take your perf measurements word for it. Thanks for trying it out though.

Reply:
@mrrodriguez @WilliamParker , digging deeper into the example i have, it would seem the time savings is not due to the logic implemented in `clojure.set`. 

The actual reason its faster is because the nodes that are being added when the number of forward forward edges is the highest, are children of the root node. Therefore the sets that are being unioned, is in actuality a single set and then goes down the single arg path of union, basically identity.

Reply:
I think that i might change to the idea that @mrrodriguez  purposed above, but special case the scenario where there is only one parent so that we are not rebuilding the set of forward-edges. 

I will try and post an update shortly.

Reply:
switching to use:
```
            forward-edges (if (= 1 (count parent-ids))
                            ;; If there is only one parent then there is no need to reconstruct
                            ;; the set of forward edges
                            (-> (:forward-edges beta-graph)
                                (get (first parent-ids)))
                            (->> (select-keys parent-ids (:forward-edges beta-graph))
                                 vals
                                 (into [] (comp cat (distinct)))))

            id-to-condition-nodes (:id-to-condition-node beta-graph)

            existing-id (some #(when (= node (get id-to-condition-nodes %))
                                 %)
                              (sort forward-edges))
```

Though this is still slightly slower than what i have currently, ~3 secs in the add-conjunctions function.